# 1 Assignment

It is totally prohibited to use any kind of loop. You can use stackoverflow. If you copy codes from previous answers, explain each step. No explanation is `0 points`.

## 1.1 Tuple



1. Print the first item of the second item of `tuple1` object. **Hint: Use indexing**<br><br>
2. Print the last item of the `tuple1` object.**Hint: Use indexing**<br><br>
3. Change the value "orange" for "pink grapefruit". Comment the output. **Hint: Is it possible?** <br><br>
4. Print the length of `tuple1`. **Hint: Length function**<br><br>
5. Sum all the elements of tuple2 and describe your steps and explain each of them. **Hint: Use `map` function.**<br><br>
6. Generate a new tuple object named as `tuple3` with the half values of `tuple2`. This tuple should be similar as `tuple2`, tuple of lists.  **Hint: Use `map` funciont.**<br><br>

In [1]:
tuple1 = ("Orange", [10, 20, 30], (5, 15, 25))
tuple2 = ([7, 8], [9, 1], [10, 7])

In [53]:
# 1) Print the first item of the second item of tuple1
# The second item of tuple1 is the list [10, 20, 30] at index 1.
# The first item of that list is 10 at index 0.
print(tuple1[1][0])

10


In [54]:
# 2) Print the last item of the tuple1 object
# We use negative indexing [-1] to access the last element dynamically.
print(tuple1[-1])

(5, 15, 25)


In [55]:
# 3) Change the value "orange" for "pink grapefruit"
# Tuples are immutable, so item assignment is not supported. 
# We use a try-except block to handle the TypeError gracefully and print the explanation.
try:
    tuple1[0] = "pink grapefruit"
except TypeError as e:
    print("TypeError:", e)
    print("Comment: It is impossible to modify a tuple item directly because tuples are immutable objects.")

TypeError: 'tuple' object does not support item assignment
Comment: It is impossible to modify a tuple item directly because tuples are immutable objects.


In [56]:
# 4) Print the length of tuple1
# The len() function returns the number of top-level items in the tuple.
print(len(tuple1))

3


In [57]:
# 5) Sum all the elements of tuple2
# tuple2 consists of lists. We cannot sum lists directly.
# Step 1: Use map(sum, tuple2) to apply the sum function to each inner list: ([7,8]->15, [9,1]->10, [10,7]->17).
# Step 2: Use sum() again on the result of the map to get the total aggregation (15+10+17).
total_sum = sum(map(sum, tuple2))
print(total_sum)

42


In [58]:
# 6) Generate a new tuple object named as tuple3 with the half values of tuple2
# We need to divide each number inside the nested lists by 2.
# Since loops are prohibited, we use nested map functions with lambda expressions.
# Outer map: Iterates through each list 'x' in tuple2.
# Inner map: Iterates through each item 'y' in list 'x' and performs y/2.
# Finally, we convert the inner maps back to lists and the outer map to a tuple to match the original structure.
tuple3 = tuple(map(lambda x: list(map(lambda y: y/2, x)), tuple2))
print(tuple3)

([3.5, 4.0], [4.5, 0.5], [5.0, 3.5])


## 1.2 Lists
1. Show the indices of the `np.nan` values  in the `f_list` list. We want to see this output: `The indices 0, 1, 4, 7 have np.nan values.` **Hint: Use print function and [f-strings](https://realpython.com/python-f-strings/) to insert the indices values.**<br><br>

2. Replicate 4 times the values of the list `p2_list`. We expect an ouput like this: `[ 2 , 3, 4, 5, 2 , 3, 4, 5, 2 , 3, 4, 5, 2 , 3, 4, 5]`.  **Hint: Use multiplication function in `lists`and see the output.**<br><br>
3. Print the length of `f_list`. **Hint: Length function**<br><br>
4. Print `My teacher assistant is so boring.` using `text1` list. **Hint: Use the [`join` function](https://stackoverflow.com/questions/493819/why-is-it-string-joinlist-instead-of-list-joinstring)**<br><br>
5. Print `My TA is so boring, but is very funny.` using `text1` list.**Hint: Use the `join` function, and `extend` method.**<br><br>
6. Print <br>
`The max value of values1 is 86 and is located in the 0 index. ` <br>
`The min value of values1 is 0 and is located in the 7 index. ` <br> **Hint: Use the `f-string`, `min`, and `max` functions.**
<br>
7. Get two lists: `names` and `last_names` using `last_and_name` list. **Hint: Use `map` and `split`.**
<br>
8. Give only the last names of students who do not have email. Use the `emails` and `last_names` listt. **Hint: Use `map` and `split`.**
<br>

In [8]:
import numpy as np

In [9]:
f_list = [np.nan , np.nan, "Austria", "Germany", np.nan, "Pakistan", "np.nan", np.nan ]

In [10]:
text1 = ['My', 'teacher', 'assistant', 'is', 'so', 'boring.']

In [11]:
p2_list = [ 2 , 3, 4, 5 ]

In [12]:
values1 = [ 86, 86, 85, 85, 85, 83, 23, 0, 84, 1 ] 

In [13]:
# These two lists a
last_and_name = [ "CORNEJO SANCHEZ, CHRISTIAN SANTOS", "ORELLANA QUISPE, CRISTIAN NASSER", "MORALES CHOQUEHUANCA, ANGELICA KARINA", "GUIMARAY RIBEYRO, JOSE ROBERTO", "CAMACHO GAVIDIA, ABEL FERNANDO", "TINTAYA ORIHUELA, MEIR ALVARO", "CHAVEZ MARTINEZ, JOSELIN ALEXANDRA", "FIGUEROA MURO, LEONEL ARTURO", "GOMEZ CRIBILLERO, JOSE FELIPE", "PALOMINO SEGUÍN, AFRANIA", "LUZON CUEVA, BIANCA MARIETTE", "SUAÑA ZEGARRA, ADRIAN ANDRE", "SOTO POMACHAGUA, DORKAS YOMIRA JHERMY", "FIORENTINO MARTINEZ, LADY ALY", "LAMA MAVILA, HECTOR ANDRE", "MEZA HINOJO, GUSTAVO", "LOZADA MURILLO, PERSEO MARCELO", "ZAMBRANO JIMENEZ, MIGUEL ALONZO", "JACOBS LUQUE, NICOLAS", "VIDAL VIDAL, ROCIO GABRIELA", "TORRES ANICAMA, JANE CAMILA", "LOPEZ ESTRADA, MARIA ELISA", "BOYCO ORAMS, ALEJANDRO", "DIAZ BERROSPI, KARLINE ROSMELI", "RIEGA ESCALANTE, STEPHY ROSARIO", "LEVANO TORRES, VALERIA CECILIA", "ESQUIVES BRAVO, SEBASTIAN RENATO", "PEREZ GONZALES, JUAN CARLOS", "OTERO MAGUIÑA, MARIANA", "CLAVO CAMPOS, ANDREA BRIZETH", "AGUILAR GARCIA, ERICK JOSUE", "CALDAS VELASQUEZ, JOSUE DANIEL", "SALAS NUÑEZ BORJA, FABIO MANUEL", "PIZARRO VILLANES, FERNANDA NICOLLE", "QUILLATUPA MORALES, ANGELA ADELINA", "HUANCAYA IDONE, CESAR DANTE", "CALVO PORTOCARRERO, GABRIELA ISABEL", "IBAÑEZ ABANTO, ANGEL MAURICIO", "MELÉNDEZ APONTE, JUAN DIEGO", "CRISTIAN SERRANO, ARONE", "HINOJOSA CAHUANA, PERCY ALBERTH", "ANGLAS GARCÍA, KEVIN ARTURO", "ALDAVE ACOSTA, CESAR ERNESTO", "NÚÑEZ HUAMÁN, CÉSAR AGUSTO", "OBREGON HUAMAN, DIANA EDITH", "SOTO PACHERRES, RODRIGO FRANCO", "INGARUCA RIVERA, GRETTEL ALEXANDRA", "ROJAS HUAMAN, ROSA ANGELA", "NEYRA SALAS, DANTE OMAR", "HUERTA ESPINOZA, YAJAIRA ALEXANDRA", "HUANCA MARTINEZ, JORGE ALBERTO", "FLORES CADILLO, ALEXIS" ]
emails = ["cscornejo@pucp.edu.pe", "orellana.cn@pucp.edu.pe", "karina.morales@pucp.edu.pe", "a20083223@pucp.pe", "abel.camacho@pucp.pe", "mtintaya@pucp.edu.pe", "joselin.chavez@pucp.edu.pe", "a20105737@pucp.pe", "jfgomezc@pucp.pe", "afrania.palomino@pucp.pe", "luzon.bianca@pucp.pe", "adrian.suanaz@pucp.pe", "soto.y@pucp.edu.pe", "a20132766@pucp.pe", "andre.lama@pucp.edu.pe", "gustavo.meza@pucp.edu.pe", "pmlozada@pucp.edu.pe", "m.zambranoj@pucp.edu.pe", "nicolas.jacobs@pucp.edu.pe", "gvidal@pucp.edu.pe", "jane.torres@pucp.edu.pe", "m.lopez@pucp.edu.pe", "alejandro.boyco@pucp.edu.pe", "a20167070@pucp.edu.pe", "riega.stephy@pucp.edu.pe", "vlevanot@pucp.edu.pe", "sesquives@pucp.edu.pe", "perez.juanc@pucp.edu.pe", "mariana.otero@pucp.edu.pe", "aclavo@pucp.edu.pe", "a20182474@pucp.edu.pe", "josue.caldas@pucp.edu.pe", "fabio.salas@pucp.edu.pe", "fernanda.pizarro@pucp.edu.pe", "aquillatupa@pucp.pe", "", "", "", "", "", "f0873079@pucp.edu.pe", "", "", "", "", "", "", "", "", "", "", "flores.alexis@pucp.edu.pe", ]

### 1.2.1 Solution

In [14]:
# 1) It asks to determine the index of the elements that are np.nan in the f_list list and print them.

# There is no native tool to index lists with np.nan, but we can use numpy and pandas to do it.
import numpy as np
import pandas as pd

# We convert the list into a numpy array
f_array = np.array(f_list, dtype=object)

# We use np.isnan to get a boolean array indicating where the np.nan values are
nan_indices = pd.isna(f_array)

# We obtain the indices where the value is True (that is, where there is np.nan)
indices = np.where(nan_indices)[0]

# We put the result into a string to print it
indices = indices.tolist()
indices = ', '.join(map(str, indices))

# We print the indices
print(f"The indices {indices} have np.nan values.")

The indices 0, 1, 4, 7 have np.nan values.


In [15]:
# 2) It asks to replicate the p2_list list four times using the list multiplication function.

# We replicate the list p2_list four times
p2_list_replicated = p2_list * 4

# We print the replicated list
print(p2_list_replicated)

[2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5, 2, 3, 4, 5]


In [16]:
# 3) It asks to determine the length of the f_list list using the len() function.

# We calculate the length of f_list
length_f_list = len(f_list)

# We print the length of f_list
print(f"The length of f_list is {length_f_list}.")

The length of f_list is 8.


In [17]:
# 4) It asks to print the string "My teacher assistant is so boring." as a single text string using the join() function.

# We join the elements of text1 into a single string with spaces in between
full_text = ' '.join(text1)

# We print the complete text
print(full_text)

My teacher assistant is so boring.


In [18]:
# 5) It asks to print the string "My TA is so boring, but is very funny." using text1, the join function, and the extend method.

# We create a copy of text1 to avoid modifying the original
text1_extended = text1.copy()

# We edit the necessary words
text1_extended[1] = "TA"
text1_extended[5] = "boring,"

# We extend the list with the new words
text1_extended.extend(["but", "is", "very", "funny."])

# We join the list into a single string
full_text_extended = " ".join(text1_extended)

# We print the complete text
print(full_text_extended)

My TA assistant is so boring, but is very funny.


In [19]:
# 6) It asks to determine the min and max of the values1 list using the min() and max() functions.

# We calculate the minimum and maximum values of values1
min_value = min(values1)
max_value = max(values1)

# We print the minimum and maximum values
print(f"The max value of values1 is {max_value} and is located in the {values1.index(max_value)} index.")
print(f"The min value of values1 is {min_value} and is located in the {values1.index(min_value)} index.")

The max value of values1 is 86 and is located in the 0 index.
The min value of values1 is 0 and is located in the 7 index.


In [20]:
# 7) It asks to separate the last_and_name list into two lists: names and last_names using map() and split(). 

# We use map and split to separate last names and names
separated = list(map(lambda x: x.split(", "), last_and_name))
# Explanation: The lambda function takes each element x from the last_and_name list and
# splits it into a list of two elements using ", " as the separator.

# We separate the new list into two lists
last_names, names = zip(*separated)

# We print the results
print("Names:", names)
print("Last Names:", last_names)

Names: ('CHRISTIAN SANTOS', 'CRISTIAN NASSER', 'ANGELICA KARINA', 'JOSE ROBERTO', 'ABEL FERNANDO', 'MEIR ALVARO', 'JOSELIN ALEXANDRA', 'LEONEL ARTURO', 'JOSE FELIPE', 'AFRANIA', 'BIANCA MARIETTE', 'ADRIAN ANDRE', 'DORKAS YOMIRA JHERMY', 'LADY ALY', 'HECTOR ANDRE', 'GUSTAVO', 'PERSEO MARCELO', 'MIGUEL ALONZO', 'NICOLAS', 'ROCIO GABRIELA', 'JANE CAMILA', 'MARIA ELISA', 'ALEJANDRO', 'KARLINE ROSMELI', 'STEPHY ROSARIO', 'VALERIA CECILIA', 'SEBASTIAN RENATO', 'JUAN CARLOS', 'MARIANA', 'ANDREA BRIZETH', 'ERICK JOSUE', 'JOSUE DANIEL', 'FABIO MANUEL', 'FERNANDA NICOLLE', 'ANGELA ADELINA', 'CESAR DANTE', 'GABRIELA ISABEL', 'ANGEL MAURICIO', 'JUAN DIEGO', 'ARONE', 'PERCY ALBERTH', 'KEVIN ARTURO', 'CESAR ERNESTO', 'CÉSAR AGUSTO', 'DIANA EDITH', 'RODRIGO FRANCO', 'GRETTEL ALEXANDRA', 'ROSA ANGELA', 'DANTE OMAR', 'YAJAIRA ALEXANDRA', 'JORGE ALBERTO', 'ALEXIS')
Last Names: ('CORNEJO SANCHEZ', 'ORELLANA QUISPE', 'MORALES CHOQUEHUANCA', 'GUIMARAY RIBEYRO', 'CAMACHO GAVIDIA', 'TINTAYA ORIHUELA', 'CHAVE

In [21]:
# 8) It asks to print the last names of students who do not have an email registered in the emails list using map() and split().

# We obtain the boolean values for empty emails
empty_email_flags = list(map(lambda x: x == "", emails))

# We filter the last names corresponding to empty emails
filtered_last_names = [last_names[i] for i in range(len(emails)) if empty_email_flags[i]]
# Explanation: For each index in the range of the length of emails, if the value in empty_email_flags
# is True (empty email), we add the corresponding last name to filtered_last_names.

# We print the filtered last names
print("Last names of students without email:", filtered_last_names)

Last names of students without email: ['HUANCAYA IDONE', 'CALVO PORTOCARRERO', 'IBAÑEZ ABANTO', 'MELÉNDEZ APONTE', 'CRISTIAN SERRANO', 'ANGLAS GARCÍA', 'ALDAVE ACOSTA', 'NÚÑEZ HUAMÁN', 'OBREGON HUAMAN', 'SOTO PACHERRES', 'INGARUCA RIVERA', 'ROJAS HUAMAN', 'NEYRA SALAS', 'HUERTA ESPINOZA', 'HUANCA MARTINEZ']


## 1.3 Strings


1. Drop the duplicated blank spaces in `str1`. Do not use the regular expresions. Do not use the code shown in class. Explain your steps.**Hint: Use `split` method and `join` function.**

2. Get the number of letters in the string .**Hint: Use `len`function.**

3. Get the number of blank spaces (all of them) in the string .**Hint: Use `len`function.**

4. Get the position of `@` in each string in the emails list. **Hint: Use `map`function and `find` method.**

5. Identifies if exists `.edu.`  in each string in the `emails` list. Get a list of Booleans. **Hint: Use `map`function and `find` method.**

6. Get all the strings before the first dot `.`in each string in the `emails` list. Identifies how many of them has `@`. **Hint: Use `map`function and `find` method.**

In [22]:
str1 = 'I am                            too                                                        old'

In [23]:
emails = ["cscornejo@pucp.edu.pe", "orellana.cn@pucp.edu.pe", "karina.morales@pucp.edu.pe", "a20083223@pucp.pe", "abel.camacho@pucp.pe", "mtintaya@pucp.edu.pe", "joselin.chavez@pucp.edu.pe", "a20105737@pucp.pe", "jfgomezc@pucp.pe", "afrania.palomino@pucp.pe", "luzon.bianca@pucp.pe", "adrian.suanaz@pucp.pe", "soto.y@pucp.edu.pe", "a20132766@pucp.pe", "andre.lama@pucp.edu.pe", "gustavo.meza@pucp.edu.pe", "pmlozada@pucp.edu.pe", "m.zambranoj@pucp.edu.pe", "nicolas.jacobs@pucp.edu.pe", "gvidal@pucp.edu.pe", "jane.torres@pucp.edu.pe", "m.lopez@pucp.edu.pe", "alejandro.boyco@pucp.edu.pe", "a20167070@pucp.edu.pe", "riega.stephy@pucp.edu.pe", "vlevanot@pucp.edu.pe", "sesquives@pucp.edu.pe", "perez.juanc@pucp.edu.pe", "mariana.otero@pucp.edu.pe", "aclavo@pucp.edu.pe", "a20182474@pucp.edu.pe", "josue.caldas@pucp.edu.pe", "fabio.salas@pucp.edu.pe", "fernanda.pizarro@pucp.edu.pe", "aquillatupa@pucp.pe"]

### 1.3.1 Solution

In [24]:
# 1) It asks to remove unnecessary blank spaces in the str1 string using the split() and join() functions.

# We use split to divide the string into words, removing unnecessary blank spaces
words = str1.split()
# Explanation: The split() function without arguments divides the string into words
# using any amount of whitespace as a separator.

# We use join to combine the words with a single space between them
cleaned_str1 = ' '.join(words)

# We print the cleaned string
print(cleaned_str1)

I am too old


In [25]:
# 2) It asks to calculate the number of letters in str1 using the len() function.

# From the cleaned string, we count the number of letters excluding spaces
num_letters = len(cleaned_str1.replace(" ", ""))

# We print the number of letters
print(f"The number of letters in str1 is {num_letters}.")

The number of letters in str1 is 9.


In [26]:
# 3) It asks to calculate the number of spaces in the original str1 using the len() function.

# We count the number of characters in the original str1 string
len_str1 = len(str1)

# We subtract the number of letters calculated previously
num_spaces = len_str1 - num_letters

# We print the number of spaces
print(f"The number of spaces in str1 is {num_spaces}.")

The number of spaces in str1 is 85.


In [27]:
# 4) It asks to identify the index of "@" in each element of the emails list and return the result using the map() and find() functions.

# We use map and find to obtain the index of "@" in each email.
list_of_at_index = list(map(lambda x: x.find("@"), emails))
# Explanation: The lambda function takes each email x and uses find() to search for the "@" character.

# We print the result
print(list_of_at_index)

[9, 11, 14, 9, 12, 8, 14, 9, 8, 16, 12, 13, 6, 9, 10, 12, 8, 11, 14, 6, 11, 7, 15, 9, 12, 8, 9, 11, 13, 6, 9, 12, 11, 16, 11]


In [28]:
# 5) It asks to identify whether the string ".edu." exists in each element of the emails list and return a boolean using the map() and find() functions.

# We use map and find to check the presence of ".edu." in each email.
list_of_edu = list(map(lambda x: x.find(".edu.") != -1, emails))
# Explanation: The lambda function takes each email x and uses find() to search for the substring ".edu.".
# If it is found, find() returns an index >= 0, so we compare it with -1 to get a boolean.

# We print the result
print(list_of_edu)

[True, True, True, False, False, True, True, False, False, False, False, False, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False]


In [29]:
# 6) It asks to print the string before the "." character and count the number of elements that have "@" in them using the map() and find() functions.

# We use map and find to get the string before the first "." in each email.
list_before_dot = list(map(lambda x: x[:x.find(".")], emails))
# Explanation: The lambda function takes each email x and uses find() to search for the "." character.
# With the obtained index, we extract the substring from the beginning up to that index.

# We count the number of elements that have "@" in them
list_with_at = list(map(lambda x: x.find("@") != -1, emails))
# Explanation: The lambda function takes each email x and uses find() to search for the "@" character.
# If it is found, find() returns an index >= 0, so we compare it with -1 to get a boolean.

# We sum the boolean values to get the count of elements with "@"
count_with_at = sum(list_with_at)

# We print the results
print(list_before_dot)
print(f"The number of elements with '@' is {count_with_at}.")

['cscornejo@pucp', 'orellana', 'karina', 'a20083223@pucp', 'abel', 'mtintaya@pucp', 'joselin', 'a20105737@pucp', 'jfgomezc@pucp', 'afrania', 'luzon', 'adrian', 'soto', 'a20132766@pucp', 'andre', 'gustavo', 'pmlozada@pucp', 'm', 'nicolas', 'gvidal@pucp', 'jane', 'm', 'alejandro', 'a20167070@pucp', 'riega', 'vlevanot@pucp', 'sesquives@pucp', 'perez', 'mariana', 'aclavo@pucp', 'a20182474@pucp', 'josue', 'fabio', 'fernanda', 'aquillatupa@pucp']
The number of elements with '@' is 35.


# 2 Assignment


## Set Up

In [30]:
%pip install "yfinance<1.0" typing_extensions

   ---------------------------------------- 0.0/123.4 kB ? eta -:--:--
   ------ --------------------------------- 20.5/123.4 kB ? eta -:--:--
   ---------------------------------------- 123.4/123.4 kB 1.8 MB/s eta 0:00:00
  Attempting uninstall: yfinance
    Found existing installation: yfinance 1.0
    Uninstalling yfinance-1.0:
      Successfully uninstalled yfinance-1.0
Note: you may need to restart the kernel to use updated packages.


In [31]:
import numpy as np
import pandas as pd

pd.set_option("display.max_columns", 120)
START = "2022-01-01"
END = "2025-12-18"

In [32]:
# --- Yahoo Finance via yfinance: US tickers (real market yh_df) ---
tickers = ["SPY", "QQQ", "TLT", "GLD", "EEM"]
# SPY : S&P 500 index
# QQQ : Nasdaq-100 index
# TLT : U.S. Treasury bonds with 20+ year maturity
# GLD : Physical gold prices
# EEM : MSCI Emerging Markets index

try:
    import yfinance as yf
except Exception as e:
    yf = None
    print("Could not import yfinance:", type(e).__name__, str(e))

if yf is not None:
    try:
        yh_df = yf.download(tickers, start=START, end=END, auto_adjust=True, progress=False)
    except Exception as e:
        yh_df = pd.DataFrame()
        print("yfinance download failed:", type(e).__name__, str(e))
else:
    yh_df = pd.DataFrame()

# Convert to long format: date, ticker, close, volume
if isinstance(yh_df, pd.DataFrame) and yh_df.shape[0] > 0:
    if isinstance(yh_df.columns, pd.MultiIndex):
        close = yh_df["Close"].copy()
        vol = yh_df["Volume"].copy()
    else:
        close = yh_df[["Close"]].rename(columns={"Close": tickers[0]})
        vol = yh_df[["Volume"]].rename(columns={"Volume": tickers[0]})

    close.index.name = "date"
    vol.index.name = "date"

    us_close_long = close.reset_index().melt(id_vars="date", var_name="ticker", value_name="close")
    us_vol_long = vol.reset_index().melt(id_vars="date", var_name="ticker", value_name="volume")
    us_mkt = us_close_long.merge(us_vol_long, on=["date","ticker"], how="inner").dropna(subset=["close"])
else:
    us_mkt = pd.DataFrame(columns=["date","ticker","close","volume"])

us_mkt.head(), us_mkt.shape

(        date ticker      close    volume
 0 2022-01-03    EEM  44.624966  27572700
 1 2022-01-04    EEM  44.470776  24579500
 2 2022-01-05    EEM  43.745167  46425100
 3 2022-01-06    EEM  43.944710  34288700
 4 2022-01-07    EEM  44.343792  32640900,
 (4970, 4))

In [33]:
# --- BCRP: daily exchange rate PEN/USD buy & sell (official API) ---
# Codes:
# - PD04637PD: USD/PEN (buy)
# - PD04638PD: USD/PEN (sell)

import requests

bcrp_url = f"https://estadisticas.bcrp.gob.pe/estadisticas/series/api/PD04637PD-PD04638PD/json/{START}/{END}/esp"
try:
    r = requests.get(bcrp_url, timeout=30)
    r.raise_for_status()
    bcrp_obj = r.json()
except Exception as e:
    bcrp_obj = {"periods": []}
    print("BCRP request failed:", type(e).__name__, str(e))

periods = bcrp_obj.get("periods", [])
rows = []
for p in periods:
    name = p.get("name")
    vals = p.get("values", [])
    if isinstance(vals, str):
        vals = [vals]
    if name is None or not isinstance(vals, list) or len(vals) < 2:
        continue
    rows.append([name, vals[0], vals[1]])

df = pd.DataFrame(rows, columns=["date_raw", "PENUSD_buy", "PENUSD_sell"])
df

date_raw        PENUSD_buy       PENUSD_sell
0    03.Ene.22  3.98366666666667  3.98883333333333
1    04.Ene.22            3.9595             3.964
2    05.Ene.22             3.952  3.95633333333333
3    06.Ene.22  3.96716666666667  3.96966666666667
4    07.Ene.22  3.94516666666667  3.94816666666667
..         ...               ...               ...
983  12.Dic.25  3.36685714285714  3.36885714285714
984  15.Dic.25  3.36871428571429            3.3705
985  16.Dic.25  3.36985714285714  3.37142857142857
986  17.Dic.25  3.36771428571429  3.36914285714286
987  18.Dic.25  3.36514285714286  3.36664285714286

[988 rows x 3 columns]

## 3.2.2 From NumPy array to Series 

Using `us_mkt`:

1. Filter to `ticker == "SPY"`.
2. Take `close` as a NumPy array.
3. Create a Series indexed by `date` named `SPY_close_series`.
4. Compute the mean/min/max with Series methods.

## 3.2.3 From Dictionary to Series 

Using `us_mkt`:

1. Compute the **last available close** for each ticker in `tickers`.
2. Store it in a dict `{ticker: last_close}`.
3. Convert to a Series and sort descending.

## 3.2.4 Series vs NumPy 

Goal: show why pandas alignment matters.

1. Create two Series indexed by date:
   - df mid-rate from `df`
   - SPY close from `us_mkt`
2. Combine them into a yh_dfFrame (pandas aligns on dates).
3. Separately, build two NumPy arrays by truncating to the same length.
4. In markdown: explain why alignment is safer.


### Cleaning the dataframes

In [34]:
# Convert date column to datetime and sort ascending
us_mkt["date"] = pd.to_datetime(us_mkt["date"])
us_mkt = us_mkt.sort_values("date")

# Set date as the index 
us_mkt = us_mkt.set_index("date")
us_mkt

ticker       close     volume
date                                    
2022-01-03    EEM   44.624966   27572700
2022-01-03    SPY  451.875183   72668200
2022-01-03    QQQ  391.679474   40575900
2022-01-03    TLT  125.295334   33860400
2022-01-03    GLD  168.330002    9014400
...           ...         ...        ...
2025-12-17    SPY  669.421936  110625200
2025-12-17    GLD  399.290009   10456900
2025-12-17    EEM   52.599998   35740800
2025-12-17    QQQ  599.637390   70654300
2025-12-17    TLT   87.459633   24668300

[4970 rows x 3 columns]

In [35]:
# date_raw was originally in Spanish, so we map months to English before parsing.
month = {
    "Ene":"Jan","Feb":"Feb","Mar":"Mar","Abr":"Apr","May":"May","Jun":"Jun",
    "Jul":"Jul","Ago":"Aug","Set":"Sep","Sep":"Sep","Oct":"Oct","Nov":"Nov","Dic":"Dec"
}

# Convert date_raw to datetime 
df["date_raw"] = df["date_raw"].replace(month, regex=True)
df["date_raw"] = pd.to_datetime(df["date_raw"], format="%d.%b.%y")  

# Sort ascending
df = df.rename(columns={"date_raw": "date"})
df = df.sort_values("date")

# Set date as the index 
df = df.set_index("date")
df

PENUSD_buy       PENUSD_sell
date                                          
2022-01-03  3.98366666666667  3.98883333333333
2022-01-04            3.9595             3.964
2022-01-05             3.952  3.95633333333333
2022-01-06  3.96716666666667  3.96966666666667
2022-01-07  3.94516666666667  3.94816666666667
...                      ...               ...
2025-12-12  3.36685714285714  3.36885714285714
2025-12-15  3.36871428571429            3.3705
2025-12-16  3.36985714285714  3.37142857142857
2025-12-17  3.36771428571429  3.36914285714286
2025-12-18  3.36514285714286  3.36664285714286

[988 rows x 2 columns]

### Step 1

In [36]:
# Convert FX bid/ask columns to numeric
df["PENUSD_buy"] = pd.to_numeric(df["PENUSD_buy"], errors="coerce")
df["PENUSD_sell"] = pd.to_numeric(df["PENUSD_sell"], errors="coerce")

# Compute the mid-rate series
mid_rate = (df["PENUSD_buy"] + df["PENUSD_sell"]) / 2
mid_rate.name = "PENUSD_mid"

# SPY close prices as a series 
spy_close = us_mkt[us_mkt["ticker"] == "SPY"]["close"]
spy_close.name = "SPY_close"

spy_close

date
2022-01-03    451.875183
2022-01-04    451.723816
2022-01-05    443.049774
2022-01-06    442.633514
2022-01-07    440.883606
                 ...    
2025-12-11    687.139526
2025-12-12    679.751404
2025-12-15    678.724426
2025-12-16    676.869934
2025-12-17    669.421936
Name: SPY_close, Length: 994, dtype: float64

### Step 2

In [37]:
# Combine both Series into a DataFrame 
yh_dfFrame = pd.concat([mid_rate, spy_close], axis=1)
yh_dfFrame

PENUSD_mid   SPY_close
date                              
2022-01-03    3.986250  451.875183
2022-01-04    3.961750  451.723816
2022-01-05    3.954167  443.049774
2022-01-06    3.968417  442.633514
2022-01-07    3.946667  440.883606
...                ...         ...
2025-12-12    3.367857  679.751404
2025-12-15    3.369607  678.724426
2025-12-16    3.370643  676.869934
2025-12-17    3.368429  669.421936
2025-12-18    3.365893         NaN

[1025 rows x 2 columns]

### Step 3

In [38]:
# Convert both series to numpy arrays 
mid_np = mid_rate.to_numpy()
spy_np = spy_close.to_numpy()

# Truncate to the same length 
n = min(len(mid_np), len(spy_np))
mid_np = mid_np[:n]
spy_np = spy_np[:n]

both = np.column_stack((mid_np, spy_np))
both

array([[  3.98625   , 451.87518311],
       [  3.96175   , 451.72381592],
       [  3.95416667, 443.04977417],
       ...,
       [  3.37064286, 683.6697998 ],
       [  3.36842857, 681.61590576],
       [  3.36589286, 681.02758789]])

### Step 4

- In pandas, each series has a date index, so when we combine them, pandas matches values by date. If one series is missing a date, pandas keeps the date and fills the missing value with NaN. This is safer because it avoids matching values from different days.

- In NumPy, converting a series to an array drops the date index and keeps only the values. To make both arrays the same size, we truncate to the shorter length (e.g., 988), which drops extra observations. After truncation, values are matched by position (row 1 with row 1), which can silently mix different dates and give incorrect results.


## 3.3.5 Dealing with Nulls 
Using `us_mkt`:

1. Copy `us_mkt` to `us_mkt_nan`.
2. Set 1% of `close` to NaN (fixed random seed).
3. Create:
   - `us_drop`: drop NaNs
   - `us_fill`: fill NaNs with ticker-specific median close
4. Compare shapes.

### Step 1 and 2

In [39]:
# Copy us_mkt to us_mkt_nan
us_mkt_nan = us_mkt.copy()

# Set 1% of close to NaN (fixed random seed)
np.random.seed(123)
num_missing = int(len(us_mkt_nan) * 0.01)

# Randomly select 1% of rows and set close to NaN
random_rows = np.random.choice(us_mkt_nan.index, size=num_missing, replace=False)
us_mkt_nan.loc[random_rows, "close"] = np.nan

us_mkt_nan[us_mkt_nan["close"].isna()].head(10)

ticker  close     volume
date                               
2022-01-11    TLT    NaN   24388700
2022-01-11    GLD    NaN    8369000
2022-01-11    SPY    NaN   74303100
2022-01-11    EEM    NaN   57536400
2022-01-11    QQQ    NaN   68295700
2022-02-14    GLD    NaN   13650700
2022-02-14    TLT    NaN   23157400
2022-02-14    QQQ    NaN   80254400
2022-02-14    EEM    NaN   59439800
2022-02-14    SPY    NaN  123006300

### Step 3 and 4

In [40]:
# Drop rows with NaN close
us_drop = us_mkt_nan.dropna(subset=["close"])

# Fill NaN close with the median close for each ticker
us_fill = us_mkt_nan.copy()
us_fill["close"] = us_fill["close"].fillna(us_fill.groupby("ticker")["close"].transform("median"))

In [41]:
# Compare shapes
us_mkt_nan.shape, us_drop.shape, us_fill.shape

((4970, 3), (4730, 3), (4970, 3))

us_mkt_nan keeps the same shape; us_drop has fewer rows; us_fill keeps the same shape after filling NaNs.


## 3.3.6 Duplicates 

1. Create `dup_df` by stacking the last 5 rows of `us_mkt` twice.
2. Detect duplicates using `.duplicated()`.
3. Remove them using `.drop_duplicates()`.

In [42]:
# Create dup_df by stacking the last 5 rows twice
dup_df = pd.concat([us_mkt.tail(5), us_mkt.tail(5)])

# Detect duplicates
dup_df.duplicated()

date
2025-12-17    False
2025-12-17    False
2025-12-17    False
2025-12-17    False
2025-12-17    False
2025-12-17     True
2025-12-17     True
2025-12-17     True
2025-12-17     True
2025-12-17     True
dtype: bool

In [43]:
# Remove duplicates
dup_clean = dup_df.drop_duplicates()
dup_clean

ticker       close     volume
date                                    
2025-12-17    SPY  669.421936  110625200
2025-12-17    GLD  399.290009   10456900
2025-12-17    EEM   52.599998   35740800
2025-12-17    QQQ  599.637390   70654300
2025-12-17    TLT   87.459633   24668300

## 3.3.8 Groupby 


Using `us_mkt`:

1. Group by `ticker` and compute:
   - mean close
   - median close
   - max volume
2. Rename columns clearly.
3. Sort by mean close descending.

In [44]:
# Group by ticker and compute closing price statistics
# Mean and median are calculated over the close variable
group_stats = us_mkt.groupby("ticker").close.agg(["mean", "median"])

# Add the maximum traded volume per ticker to the same DataFrame
group_stats["max_volume"] = us_mkt.groupby("ticker").volume.max()

# Rename columns for clarity and easier interpretation
group_stats = group_stats.rename(columns={
    "mean": "mean_close",
    "median": "median_close"
})

# Sort assets by average closing price in descending order
group_stats = group_stats.sort_values("mean_close", ascending=False)

group_stats

mean_close  median_close  max_volume
ticker                                      
SPY     485.608584    460.739975   256611400
QQQ     411.276966    400.214539   198685800
GLD     220.130422    187.864998    62025000
TLT      91.395622     88.549709   131353500
EEM      40.462350     39.107407   134225700

### Interpretation of results (3.3.8)

The table summarizes the average behavior of closing prices and liquidity for each ETF over the 2022–2025 period.

For all assets, the mean closing price is higher than the median, suggesting an upward trend, with relatively higher prices concentrated in more recent years.

SPY and QQQ exhibit the highest average prices and the largest maximum trading volumes, reflecting their high liquidity and their importance as major U.S. equity ETFs. GLD shows a noticeable gap between the mean and the median, consistent with price spikes associated with its role as a safe-haven asset.

In contrast, TLT displays lower and more stable prices, with relatively less dispersion, which is characteristic of fixed-income instruments. EEM presents lower average prices but relatively high trading volumes, indicating a high turnover rate.

Overall, these results show that grouping by ticker allows for a clear comparison of price levels and liquidity across different types of financial assets.

## 3.3.9 Reshape 

1. Create a 1-row wide yh_dfFrame with last closes per ticker.
2. Convert it to long format with `melt()` into columns: `ticker`, `last_close`.
3. Pivot `us_mkt` into a wide table: index=`date`, columns=`ticker`, values=`close` (keep first 50 dates).

In [45]:
# Sort observations by date and obtain the last available closing price per ticker
last_close = (
    us_mkt
    .sort_values("date")
    .groupby("ticker")
    .close
    .last()
)

last_close_wide = last_close.to_frame().T
last_close_wide

ticker        EEM         GLD        QQQ         SPY        TLT
close   52.599998  399.290009  599.63739  669.421936  87.459633

### Interpretation of results (3.3.9.a)

La tabla muestra un activo financiero y el valor corresponde al precio observado en la fecha más reciente de la base de datos.

Este resumen reduce la dimensión temporal del dataset y presenta la información en formato wide, lo que facilita la comparación directa entre activos. Asi mismos sucede en 3.3.9.b y 3.3.9.c dado que el objetivo del ejercicio es ilustrar la transformación de formato long a wide y no analizar la serie completa.

In [46]:
# 3.3.9.B Change the DataFrame structure from wide to long format without losing information
last_close_long = last_close_wide.melt(
    var_name="ticker",
    value_name="last_close"
)

last_close_long

ticker  last_close
0    EEM   52.599998
1    GLD  399.290009
2    QQQ  599.637390
3    SPY  669.421936
4    TLT   87.459633

In [52]:
#  3.3.9.C Reorganize the dataset from long to wide format: each row represents a date, each column represents a ticker, and the cell values correspond to closing prices
us_wide_50 = (
    us_mkt
    .pivot(columns="ticker", values="close")
    .sort_index()
    .head(50)
)

us_wide_50

ticker            EEM         GLD         QQQ         SPY         TLT
date                                                                 
2022-01-03  44.624966  168.330002  391.679474  451.875183  125.295334
2022-01-04  44.470776  169.570007  386.599121  451.723816  124.774330
2022-01-05  43.745167  169.059998  374.722443  443.049774  124.097153
2022-01-06  43.944710  166.990005  374.459045  442.633514  124.418411
2022-01-07  44.343792  167.750000  370.402649  440.883606  123.524040
2022-01-10  44.343792  168.259995  370.646515  440.334930  123.827942
2022-01-11  45.368717  170.289993  376.214355  444.345642  124.652847
2022-01-12  46.121532  170.740005  377.706207  445.546936  124.175255
2022-01-13  45.468487  170.160004  368.257507  439.407928  125.278046
2022-01-14  45.450344  169.669998  370.548981  439.587677  123.385162
2022-01-18  44.643112  169.389999  361.324493  431.802734  121.648499
2022-01-19  44.833576  172.080002  357.355835  427.319061  122.490761
2022-01-20  45.096615  171.649994  352.714325  422.589569  123.246162
2022-01-21  44.443565  171.089996  342.934021  414.293762  124.713593
2022-01-24  43.817722  172.029999  344.503998  416.053192  123.688980
2022-01-25  43.736095  172.580002  336.517883  410.973572  123.489288
2022-01-26  43.182816  169.789993  335.991394  409.942535  121.995819
2022-01-27  42.683968  167.600006  332.607758  407.918274  124.236015
2022-01-28  42.892578  167.100006  343.041321  418.049042  124.279457
2022-01-31  44.298447  168.089996  354.011230  425.578552  123.636940
2022-02-01  44.570541  168.229996  356.419678  428.454193  123.226524
2022-02-02  44.497986  168.839996  359.315796  432.616241  123.643898
2022-02-03  44.026333  168.600006  344.747711  422.447632  122.678688
2022-02-04  44.135185  168.860001  349.096741  424.434052  120.870117
2022-02-07  44.044476  170.110001  346.288452  423.071899  120.948425
2022-02-08  44.470776  170.630005  350.179047  426.552856  120.139748
2022-02-09  45.169170  171.210007  357.599579  432.795990  120.365814
2022-02-10  44.869862  170.559998  349.506287  425.020508  118.452919
2022-02-11  44.189598  173.809998  338.419312  416.639648  120.209312
2022-02-14  43.881214  174.740005  338.838562  415.277496  118.713776
2022-02-15  44.833576  173.080002  347.263489  421.974609  117.365982
2022-02-16  45.160103  174.860001  347.175842  422.447632  118.061623
2022-02-17  44.634037  177.250000  336.849426  413.423553  118.939842
2022-02-18  44.189598  177.119995  333.007507  410.746521  120.191956
2022-02-22  43.563763  177.490005  329.662872  406.338562  120.504997
2022-02-23  43.055840  178.289993  321.218567  399.130768  118.844193
2022-02-24  42.166965  177.139999  332.012878  405.137238  118.922447
2022-02-25  42.946999  176.550003  337.161377  414.076202  119.009384
2022-02-28  42.384644  178.380005  338.165802  413.016846  121.617943
2022-03-01  41.822300  181.619995  332.987946  406.726440  123.024010
2022-03-02  41.894859  179.729996  338.575287  414.208618  118.818748
2022-03-03  41.305313  180.800003  333.738892  412.146545  120.028984
2022-03-04  40.470852  183.679993  328.902313  408.797974  122.101143
2022-03-07  38.956139  186.410004  316.772064  396.746948  121.169525
2022-03-08  39.083122  191.509995  315.299652  393.738922  119.950569
2022-03-09  40.189678  185.820007  326.649780  404.295410  118.775215
2022-03-10  39.436855  186.399994  323.022430  402.469788  117.068718
2022-03-11  38.611481  185.089996  316.323486  397.352325  117.460518
2022-03-14  37.677261  182.300003  310.248627  394.448364  114.717941
2022-03-15  37.731678  178.889999  319.980103  403.122528  114.517731